gateway-discovery-cm/src/main/java/org/apache/knox/gateway/topology/discovery/cm/monitor/PollingConfigurationAnalyzer.java

```diff
@@ -365,12 +375,32 @@ private DiscoveryApiClient getApiClient(final ServiceDiscoveryConfig discoveryCo
                                                clusterName,
                                                lastTimestamp);
     for (ApiEvent event : events) {
-      restartEvents.add(new RestartEvent(event));
+      List<ApiEventAttribute> attributes = event.getAttributes();
+      Map<String,Object> map = getAttributeMap(attributes);
+      addIfRelevantEvent(restartEvents, event, map);
     }
 
     return restartEvents;
   }
 
+  @SuppressWarnings("unchecked")
+  private void addIfRelevantEvent(List<RestartEvent> restartEvents, ApiEvent event, Map<String, Object> map) {
+    String command = null;
+    String status = null;
+    command = (String) ((List<String>) map.get(COMMAND)).get(0);
+    status = (String) ((List<String>) map.get(COMMAND_STATUS)).get(0);
+    if (START_COMMAND.equals(command) || RESTART_COMMAND.equals(command) &&
+        SUCCEEDED_STATUS.equals(status) || STARTED_STATUS.equals(status)) {
+      restartEvents.add(new RestartEvent(event));
+    }
+  }
+
+  private Map<String, Object> getAttributeMap(List<ApiEventAttribute> attributes) {
+    Map<String,Object> map = new HashMap<>();
+    attributes.forEach(attr -> { map.put(attr.getName(), attr.getValues());});
```

Comment:
This will overwrite in the map correct? The attribute get name - is that unique? I think this is by COMMAND or COMMAND_STATUS based on the logic for eventually checking the map. If there are duplicate commands won't they get overwritten in the map? So its last write wins?

If you do a Start of SERVICEA and then Start of SERVICEB - will you see both events or just collapse it down to one?

Another example is doing start of same service twice with different results. start success then (implicit stop) then start failure. Will that get picked up?

I think the map is totally breaking the uniqueness and ordering of the events.

Reply:
The map is associated with a specific ApiEvent, so I'm not sure that the map is the concern. We may want to sort the ApiEvent collection by _timeOccurred_ to avoid issues associated with what you're describing.

Reply:
Yes, the map is redone for each event purposely. It is only a map of the attributes for the given event. The order of the events hasn't changed as a result of this patch and is driven by the REST API response and has nothing to do with the creation of this map. It works as intended.

Reply:
I think the latest revision will make the above more clear, Kevin. Thanks for the review!